### PR TITLE
feat: global secrets for pre/post cd

### LINKED ISSUE #3093: Feature: global secrets in pre/post cd
### 🔖 Feature description

Right now devtron's global secrets and config maps are available only in pre/post CI. Support for injecting config maps and secrets in global config map and secret should also be provided. These secrets should not be available if "run in application environment is selected"

### 🎤 Pitch

This feature will enable user to use global secrets for pre/post CD.

### ✌️ Solution

Just like pre/post CI, argo workflow templates can be used for injecting secrets into pre/post CI.

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by iamayushm**
# Description

Adding global secrets in pre/post CD. 
Fixes #3093 

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/restHandler/GlobalCMCSRestHandler.go [hunk 1]
```diff
@@ -77,3 +83,123 @@ func (handler *GlobalCMCSRestHandlerImpl) CreateGlobalCMCSConfig(w http.Response
...
+	err = handler.validator.Struct(bean)
+	if err != nil {
+		handler.logger.Errorw("validation err, CreateGlobalCMCSConfig", "err", err, "payload", bean)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+
+	// RBAC enforcer applying
+	token := r.Header.Get("token")
+	if ok := handler.enforcer.Enforce(token, casbin.ResourceGlobal, casbin.ActionCreate, "*"); !ok {
```
Comment (vikramdevtron) [inline api/restHandler/GlobalCMCSRestHandler.go:112]: ActionUpdate

File: api/restHandler/GlobalCMCSRestHandler.go [hunk 2]
```diff
@@ -77,3 +83,123 @@ func (handler *GlobalCMCSRestHandlerImpl) CreateGlobalCMCSConfig(w http.Response
...
+		return
+	}
+
+	userId, err := handler.userAuthService.GetLoggedInUser(r)
+	if userId == 0 || err != nil {
+		common.WriteJsonResp(w, err, "Unauthorized User", http.StatusUnauthorized)
+		return
+	}
+	token := r.Header.Get("token")
+	if ok := handler.enforcer.Enforce(token, casbin.ResourceGlobal, casbin.ActionCreate, "*"); !ok {
```
Comment (vikramdevtron) [inline api/restHandler/GlobalCMCSRestHandler.go:194]: ActionDelete

File: internal/sql/repository/GlobalCMCSRepository.go [hunk 3]
```diff
@@ -21,6 +24,12 @@ const (
 	VOLUME_CONFIG      = "volume"
 )
 
+const (
```
Comment (vikramdevtron) [inline internal/sql/repository/GlobalCMCSRepository.go:27]: if we have similar type of constants plz use them or put them in common

File: pkg/pipeline/CdWorkflowService.go [hunk 4]
```diff
@@ -144,8 +154,28 @@ func (impl *CdWorkflowServiceImpl) SubmitWorkflow(workflowRequest *CdWorkflowReq
...
-	var steps []v1alpha1.ParallelSteps
+	entryPoint := CD_WORKFLOW_NAME
+
+	globalCmCsConfigs, err := impl.globalCMCSService.FindAllActiveByPipelineType(repository2.PIPELINE_TYPE_CD)
+	if err != nil {
+		impl.Logger.Errorw("error in getting all global cm/cs config", "err", err)
+		return nil, err
+	}
+	if len(globalCmCsConfigs) > 0 {
+		entryPoint = "cd-stages-with-env"
```
Comment (vikramdevtron) [inline pkg/pipeline/CdWorkflowService.go:165]: make this entry point Constant

File: pkg/pipeline/CdWorkflowService.go [hunk 5]
```diff
@@ -144,8 +154,28 @@ func (impl *CdWorkflowServiceImpl) SubmitWorkflow(workflowRequest *CdWorkflowReq
...
+	}
+	for i := range globalCmCsConfigs {
+		globalCmCsConfigs[i].Name = globalCmCsConfigs[i].Name + "-" + strconv.Itoa(workflowRequest.WorkflowRunnerId) + "-" + CD_WORKFLOW_NAME
```
Comment (vikramdevtron) [inline pkg/pipeline/CdWorkflowService.go:168]: use Sprintf for concatenation

File: pkg/pipeline/GlobalCMCSService.go [hunk 6]
```diff
@@ -34,43 +43,63 @@ type GlobalCMCSDto struct {
 	Name       string `json:"name"  validate:"required"`
 	Type       string `json:"type" validate:"oneof=environment volume"`
 	//map of key:value, example: '{ "a" : "b", "c" : "d"}'
```
Comment (vikramdevtron) [inline pkg/pipeline/GlobalCMCSService.go:45]: remove commented line

File: internal/sql/repository/GlobalCMCSRepository.go (hunk 3 above)
Comment (iamayushm) [reply internal/sql/repository/GlobalCMCSRepository.go:27]: I am unable to find similar constant. Constant with value "CI_PIPELINE" but constant with value "CI" doesn't exist.

File: pkg/pipeline/GlobalCMCSService.go (hunk 6 above)
Comment (iamayushm) [reply pkg/pipeline/GlobalCMCSService.go:45]: this comment gives user idea about sample values which api requires.

**Review by vikramdevtron — APPROVED** [review]

**Review by vikramdevtron — APPROVED** [review]

**Pull request merged**